[NGT] Implement new seeding module for HLT Standalone Muon seeding and streamline HLT L3 Tracker Muon reconstruction

Parsifal-2045 opened the pull request:
#### PR description:

This PR implements a new Standalone (L2) Muon seeding strategy that fully utilizes the L1 Tracker Muons (L1TkMu) information. 
L3 Tracker Muon reconstruction is also streamlined, also taking advantage of the improved L1TkMu, to reduce redundancy. 
The full scope of the changes was described in the [HLT Upgrade meeting on 8/10/2024](https://indico.cern.ch/event/1460055/contributions/6147459/attachments/2941712/5168585/2024_10_08_thesis_results.pdf), while the details on the integration were presented in the [meeting on 3/12/2024](https://indico.cern.ch/event/1482695/contributions/6261220/attachments/2978788/5244435/2024-12-03%20Phase%202%20validation%20and%20integration%20in%20CMSSW.pdf#page=4).

The new L2 seeder directly matches the stubs used to produce each L1TkMu with segments in DTs/CSCs. All matched segments are assigned to the seed, together with the pT measurement from the L1TkMu. This approach produces a single collection of seeds. 
On the other hand, the current implementation produces two collections of seeds (from L1TkMu and "offline") that require matching and produces roughly a factor of 5 more seeds in ZMM simulated events with 200 PU.
The changes implemented in the seeding module results in a significant improvement in timing (about 42%) with mostly compatible efficiency.

The streamlining of the L3 reconstruction hinges on the idea of performing either the Inside-Out or the Outside-In L3 Tracker Muon reconstruction first, to then execute a second pass only for candidates that require it (i.e. either they were not reconstructed in the first pass or the quality of the track was not good enough). This approach starts from either reconstruction path (flexible in case of future changes or detectors' requirements) to produce a filtered collection of L3 tracks: the filter uses some of the same criteria used in the HLT Muon ID step (e.g. number of hits in the pixel, number of hits in the tracker, normalized chi2, ...). The filtered L3 Tracks are geometrically matched to either L1TkMu (Outside-In first) or Standalone Muons (Inside-Out first):
- if the match is successful, the track is considered good enough and passes to the next phase;
- if the match is unsuccessful, the lower-level object will be reused to seed the second reconstruction path.
From the current results, the Inside-Out first approach seems to perform better, with more than 80% of the events in a ZMM sample at 200 PU requiring a single reconstruction path, to be compared with about 20% for the Outside-In first approach.
In any case, the streamlined L3 reconstruction reduces the fakerate in the merged (L3 IO + L3 OI) tracks collection as well as timing, while maintaining compatible physics perfomance until and including the final HLT Muon ID objects. All results, validation plots and ntuples are available [here](https://lferragi.web.cern.ch/plots/muon_hlt_phase2/14_2_0_pre4_PR/)

The changes have been implemented using two new process modifiers:
- phase2L2AndL3Muons implements the new Standalone Muon seeding and executes the L3 Tracker Muon reconstruction Inside-Out first;
- phase2L3MuonsOIFirst to be used on top of the previous one, swaps the L3 Tracker Muon reconstruction from Inside-Out first to Outside-In first

These process modifiers have been included in a set of workflows targeting SingleMu, TTbar, and ZMM samples:
- .777 includes the phase2L2AndL3Muons 
- .778 includes the phase2L2AndL3Muons and the phase2L3MuonsOIFirst process modifiers
 
#### PR validation:

Validation builds on top of #46860 introducing specific validation paths for the new objects (e.g. filtered collections, objects to reuse). Physics performance mostly compatible with the current implementation. Tested on CMSSW_14_2_0_pre3 RelVals (TTbar and ZMM) (not pre4 due to those being limited to 35 PU) as well as private production with workflows 29834(.0, .777, .778) (TTbar) and 29850 (.0, .777, .778) (ZMM). 
**Please test on Phase 2 workflows using also .777 and .778 modifiers**

Parsifal-2045 committed:
Implement new Phase 2 Standalone Muon seeder
- Takes full advantage of the Phase 2 L1 Tracker Muon information
- Direct match between L1 stubs and segments in DTs/CSCs
- Extract stubs from L1Tk Muon, only checks segments in the same chamber as the stub to match
- Match based on dphi, dtheta and number of hits in the segment
- In the barrel, if no match is found in a given station, a rough extrapolation is attempted from the closest station with a match
- At most one seed per L1Tk Muon produced in a single seeding step, with the possibility to extend to exotic signatures

Parsifal-2045 committed:
Streamline HLT L3 Muon reconstruction
- execute either Inside-Out or Outside-In reconstruction first, resorting to the second pass only for candidates that require it
- Tracks filtered by quality using some of the same criteria as HLT Muon ID
- L1/L2 objects matched with bad L3 tracks are reused to seed the second reconstruction step
- Changes implemented via 2 procModifiers for IO first (default) and OI first reconstruction
- Implemented changes in SingleMu, TTbar, and ZMM workflows with .777 and .778 subfixes for IO and OI first reconstruction, respectively
- Change 2026 to Run4 in Phase 2 workflows

Parsifal-2045 committed:
Implement validation for new Phase 2 muon workflows

Parsifal-2045 committed:
Formatting muon sequences

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"3b4cc572da58f8dd3872340dd7291eb496c090ca":{"files":["Configuration/ProcessModifiers/python/phase2L2AndL3Muons_cff.py","DataFormats/MuonSeed/interface/L2MuonTrajectorySeed.h","DataFormats/MuonSeed/src/L2MuonTrajectorySeed.cc","DataFormats/MuonSeed/src/classes_def.xml","HLTrigger/Configuration/python/HLT_75e33/modules/hltL2MuonSeedsFromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTL2MuonsFromL1TkSequence_cfi.py","RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc","RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.h"],"squashed":false,"time":1733394641},"47c2e128bd5f080ecb8713037a7464a713fd6baa":{"files":["Configuration/EventContent/python/EventContent_cff.py","Validation/RecoMuon/python/associators_cff.py","Validation/RecoMuon/python/muonValidationHLT_cff.py"],"squashed":false,"time":1733394648},"81a84be4aab024b89967d0ef5f0e04a855c88857":{"files":["Configuration/ProcessModifiers/python/phase2L3MuonsOIFirst_cff.py","Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltIter2Phase2L3FromL1TkMuonPixelSeedsFiltered_cfi.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3FromL1TkMuonPixelTracksTrackingRegions_cfi.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3MuonFilter_cfi.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3MuonMerged_cfi.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3OISeedsFromL2Muons_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTMuonsSequence_cfi.py","RecoMuon/L3TrackFinder/interface/phase2HLTMuonSelectorForL3.h","RecoMuon/L3TrackFinder/src/phase2HLTMuonSelectorForL3.cc"],"squashed":false,"time":1733394646},"a979b05924710adfd93e1e93572feb6153f5ca47":{"files":["HLTrigger/Configuration/python/HLT_75e33/paths/HLT_AK4PFPuppiJet520_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_DoubleMediumChargedIsoPFTauHPS40_eta2p1_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_DoubleMediumDeepTauPFTauHPS35_eta2p1_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepCSV_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepFlavour_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_IsoMu20_eta2p1_LooseDeepTauPFTauHPS27_eta2p1_CrossL1_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_IsoMu24_FromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_Mu17_TrkIsoVVL_Mu8_TrkIsoVVL_DZ_FromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_Mu37_Mu27_FromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_Mu50_FromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepCSV_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepFlavour_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepCSV_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepFlavour_2p4_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFPuppiHT1070_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_PFPuppiMETTypeOne140_PFPuppiMHT140_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/HLT_TriMu_10_5_5_DZ_FromL1TkMuon_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/MC_BTV_cfi.py","HLTrigger/Configuration/python/HLT_75e33/paths/MC_JME_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTIter0Phase2L3FromL1TkSequence_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTIter2Phase2L3FromL1TkSequence_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTPhase2L3FromL1TkSequence_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTPhase2L3MuonsSequence_cfi.py","HLTrigger/Configuration/python/HLT_75e33/sequences/HLTPhase2L3OISequence_cfi.py"],"squashed":false,"time":1733394651}},"emoji":{},"last_seen_sha":"a979b05924710adfd93e1e93572feb6153f5ca47","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46897/42945

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #46453, #46853

mmusich commented:
test parameters:
   * workflow_opts= -w upgrade
   * workflow = 29834.0, 29834.777, 29834.778,  29850.0, 29850.777, 29850.778
   * enable = hlt_p2_integration, hlt_p2_timing

mmusich commented on HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3MuonFilter_cfi.py:
```diff
@@ -0,0 +1,25 @@
+
+hltPhase2L3MuonFilter = cms.EDProducer(
+    "phase2HLTMuonSelectorForL3",
+    l1TkMuons=cms.InputTag("l1tTkMuonsGmt"),
```
here as well put the `=` signs between spaces.

mmusich reviewed (commented):
These are just cosmetic comments from a quick transversal look. 
I'll go through the dense C++ changes in the coming days. 

A general request for clarification. 
What is the plan to re-absorb `phase2L3MuonsOIFirst` and `phase2L2AndL3Muons` process modifiers in the mainstream configuration? 

The question is not really for the PR proponent but I have a feeling we're diverging with the options for the phase-2 menu. @rovere @VourMa

Parsifal-2045 commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,842 @@
+      double deltaTheta = std::abs(segPos.theta() - 2 * std::atan(std::exp(-l1TkMuRef->phEta())));
+      LogDebug(metname) << "deltaTheta: " << deltaTheta;
+
+      if (deltaPhi > matchingPhiWindow_ or deltaTheta > 0.4) {
```
Rough check in delta theta, would it be better if it were something like `4 * matchingThetaWindow_`?

mmusich replied:
I was more wondering where the 0.4 comes from and if it should be a named constants.

Parsifal-2045 replied:
The module looks for a match by checking, in order, delta phi, hit multiplicity and delta theta. The checks following the first are basically used to pick the highest-quality match, but I've resorted to have a rough check in eta anyways to avoid bad matches if nothing better is found. Since the current theta window is set to 0.1, for the rough check I've decided to go for something quite a bit larger (0.4 in this case), but there's no specific reason for that particular value

mmusich replied:
As I see 0.4 repeated several times throughout the code of this class, I would prefer if you could make it a named constant and put a comment about it for later maintainability.

Moanwar commented on DataFormats/MuonSeed/src/classes_def.xml:
```diff
@@ -1,5 +1,6 @@
 <lcgdict>
-      <class name="L2MuonTrajectorySeed"  ClassVersion="12">
+      <class name="L2MuonTrajectorySeed"  ClassVersion="13">
+       <version ClassVersion="13" checksum="2121764882"/>
```
I just noticed that the class L2MuonTrajectorySeed is updated to ClassVersion="13", but L3MuonTrajectorySeed is not. Is that intentional?

Parsifal-2045 replied:
Yes, the `L2MuonTrajectorySeed` class was updated to contain a reference to the L1Tk Muon used to produce the seed (and the corresponding constructor), while this is not the case for  `L3MuonTrajectorySeed`

mmusich commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,842 @@
+      mat[1][1] = 0.05 * 0.05;                                                     // sigma^2(lambda)
+      mat[2][2] = 0.2 * 0.2;                                                       // sigma^2(phi)
+      mat[3][3] = 20. * 20.;                                                       // sigma^2(x_transverse)
+      mat[4][4] = 20. * 20.;                                                       // sigma^2(y_transverse)
```
where are all these numbers coming from?

Parsifal-2045 replied:
The propagation to the various muon stations have remained basically unchanged from the [current L2 MuonSeedGenerator](https://github.com/cms-sw/cmssw/blob/87f3f24501aefe75d4e0be6f3f624e890b8a621d/RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGenerator.cc#L241) (seeded by L1Tk Muons)

Parsifal-2045 replied:
The aim is for this new module to replace the current "legacy" seeding strategy. However, it's currently under a process modifier as a soft introduction while some slight inefficiencies are ironed out and its performance is assessed thoroughly.  In any case, in the end, there should be a single Standalone Muon seed from L1Tk Muons generator

mmusich commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,842 @@
+
+// In DT station 4 the top and bottom sectors are made of two chambers
+// due to material requirements. Online is not split:
+// Online sector 4 == offline sector 4 or 10, Online sector 10 == offline sector 10 or 14
```
4 and 10 better be named constants.

mmusich commented on HLTrigger/Configuration/python/HLT_75e33/modules/hltIter2Phase2L3FromL1TkMuonPixelSeedsFiltered_cfi.py:
```diff
@@ -33,3 +33,8 @@
+
+from Configuration.ProcessModifiers.phase2L2AndL3Muons_cff import phase2L2AndL3Muons
+from Configuration.ProcessModifiers.phase2L3MuonsOIFirst_cff import phase2L3MuonsOIFirst
+(phase2L2AndL3Muons & phase2L3MuonsOIFirst).toModify(hltIter2Phase2L3FromL1TkMuonPixelSeedsFiltered, L1TkMu = cms.InputTag("hltPhase2L3MuonFilter", "L1TkMuToReuse"))
```
```suggestion
(phase2L2AndL3Muons & phase2L3MuonsOIFirst).toModify(hltIter2Phase2L3FromL1TkMuonPixelSeedsFiltered, L1TkMu = "hltPhase2L3MuonFilter:L1TkMuToReuse")
```

should work.

mmusich commented on HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2L3FromL1TkMuonPixelTracksTrackingRegions_cfi.py:
```diff
@@ -22,3 +22,7 @@
+
+from Configuration.ProcessModifiers.phase2L2AndL3Muons_cff import phase2L2AndL3Muons
+from Configuration.ProcessModifiers.phase2L3MuonsOIFirst_cff import phase2L3MuonsOIFirst
+(phase2L2AndL3Muons & phase2L3MuonsOIFirst).toModify(hltPhase2L3FromL1TkMuonPixelTracksTrackingRegions.RegionPSet, input = cms.InputTag("hltPhase2L3MuonFilter", "L1TkMuToReuse"))
```
```suggestion
(phase2L2AndL3Muons & phase2L3MuonsOIFirst).toModify(hltPhase2L3FromL1TkMuonPixelTracksTrackingRegions.RegionPSet, input = "hltPhase2L3MuonFilter:L1TkMuToReuse")
```

should work.

mmusich commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,842 @@
+      continue;
+    }
+    float eta = l1TkMuRef->phEta();
+    float theta = 2 * std::atan(std::exp(-eta));
```
this pattern together with the usage of `ThreeVector` from `CLHEP` smells.
What about something possibly more efficient reducing the usage of trig functions along the lines of https://github.com/cms-sw/cmssw/commit/6272d52863ba88908a84ad05ab1de072450a29d0 ?

Parsifal-2045 replied:
The latest commit should address this. It's still computing theta (only once per L1TkMu) since it's used for the stub-segment matching, but everything has been refactored to result in fewer overall "taxing" computations

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46897/42955

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #46453, #46853

mmusich commented:
@cmsbuild, please test

Parsifal-2045 commented:
The latest commit has implemented most of the cosmetic changes, I'm now going through the actual code modifications

cmsbuild commented:
+1

**Size**: This PR adds an extra 56KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56c0a9/43348/summary.html
**COMMIT**: 4762785992908b7fffe9788f1ee3f85cf11ba8f7
**CMSSW**: CMSSW_15_0_X_2024-12-09-2300/el8_amd64_gcc12
**Additional Tests**: HLT_P2_INTEGRATION,HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46897/43348/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2024-12-09-2300&dataset=CMSSW_15_0_X_2024-12-09-2300-cmssw46897/el8_amd64_gcc12/59/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-09-2300+56c0a9/66047/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 451](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-09-2300+56c0a9/66047/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509546
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-09-2300+56c0a9/66047/triggerResults): found differences in 1 / 44 workflows

rovere commented:
type: ngt

rovere commented:
type ngt

Parsifal-2045 committed:
Address PR comments
- Introduce more named constants for later maintainability
- Make the maxChi2 in the estimator customizable
- Use vdt trig functions and precompute all values when possible
- Fix code style: remove not-needed member variable, use DTChamberId
  constants instead of magic numbers

mmusich reviewed (commented):
While you fix the code quality checks

Parsifal-2045 commented on RecoMuon/L3TrackFinder/interface/phase2HLTMuonSelectorForL3.h:
```diff
@@ -0,0 +1,81 @@
+  class EventSetup;
+}  // End namespace edm
+
+class phase2HLTMuonSelectorForL3 : public edm::stream::EDProducer<> {
```
Should have addressed both the quality checks (weird newline) and this

cmsbuild commented:
Pull request #46897 was updated. @AdrianoDee, @Martin-Grunewald, @Moanwar, @antoniovagnerini, @antoniovilela, @cmsbuild, @davidlange6, @dickychant, @fabiocos, @jfernan2, @mandrenguyen, @miquork, @mmusich, @rappoccio, @rseidita, @srimanob, @subirsarkar can you please check and sign again.

felicepantaleo commented:
allow @Parsifal-2045  test rights

Parsifal-2045 commented:
@cmsbuild, please test

mmusich commented:
> The new L2 seeder directly matches the stubs used to produce each L1TkMu with segments in DTs/CSCs.

when I try to run the phase-2 integration tests with the command `hltPhase2UpgradeIntegrationTests --procModifiers phase2L2AndL3Muons` I am getting crashes of the kind:

```
----- Begin Fatal Exception 11-Dec-2024 15:14:38 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 4 stream: 0
   [1] Running path 'HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepFlavour_2p4'
   [2] Calling method for module Phase2L2MuonSeedCreator/'hltL2MuonSeedsFromL1TkMuon'
Exception Message:
RefCore: A request to resolve a reference to a product of type 'std::vector<l1t::MuonStub>' with ProductID '1:379'
can not be satisfied because the product cannot be found.
Probably the branch containing the product is not stored in the input file.
   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```

this means that at the moment we're not saving in the phase2  `FEVTDEBUG ` data-tier the needed collection for the seeding after running the steps `L1TrackTrigger,L1,L1P2GT,DIGI2RAW` .

The tests generates the RAW data on the fly via

```
cmsDriver.py TTbar_14TeV_TuneCP5_cfi -s GEN,SIM,DIGI:pdigi_valid,L1TrackTrigger,L1,L1P2GT,DIGI2RAW -n 10 --conditions auto:phase2_realistic_T33 --beamspot DBrealisticHLLHC --datatier GEN-SIM-DIGI-RAW --eventcontent FEVTDEBUG --geometry ExtendedRun4D110 --era Phase2C17I13M9 --relval 9000,100 --fileout hlt_test_configs/step1.root --nThreads 1 --customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000 --python_filename hlt_test_configs/TTbar_GEN_SIM_step.py
```

should the process modifier trigger the saving of the collection in the `FEVTDEBUG` content as well?

Parsifal-2045 commented:
> > The new L2 seeder directly matches the stubs used to produce each L1TkMu with segments in DTs/CSCs.
> 
> when I try to run the phase-2 integration tests with the command `hltPhase2UpgradeIntegrationTests --procModifiers phase2L2AndL3Muons` I am getting crashes of the kind:
> 
> ```
> ----- Begin Fatal Exception 11-Dec-2024 15:14:38 CET-----------------------
> An exception of category 'ProductNotFound' occurred while
>    [0] Processing  Event run: 1 lumi: 1 event: 4 stream: 0
>    [1] Running path 'HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepFlavour_2p4'
>    [2] Calling method for module Phase2L2MuonSeedCreator/'hltL2MuonSeedsFromL1TkMuon'
> Exception Message:
> RefCore: A request to resolve a reference to a product of type 'std::vector<l1t::MuonStub>' with ProductID '1:379'
> can not be satisfied because the product cannot be found.
> Probably the branch containing the product is not stored in the input file.
>    Additional Info:
>       [a] If you wish to continue processing events after a ProductNotFound exception,
> add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.
> 
> ----- End Fatal Exception -------------------------------------------------
> ```
> 
> this means that at the moment we're not saving in the phase2 `GEN-SIM-DIGI-RAW` data-tier the needed collection for the seeding after running the steps `L1TrackTrigger,L1,L1P2GT,DIGI2RAW` .
> 
> The tests generates the RAW data on the fly via
> 
> ```
> cmsDriver.py TTbar_14TeV_TuneCP5_cfi -s GEN,SIM,DIGI:pdigi_valid,L1TrackTrigger,L1,L1P2GT,DIGI2RAW -n 10 --conditions auto:phase2_realistic_T33 --beamspot DBrealisticHLLHC --datatier GEN-SIM-DIGI-RAW --eventcontent FEVTDEBUG --geometry ExtendedRun4D110 --era Phase2C17I13M9 --relval 9000,100 --fileout hlt_test_configs/step1.root --nThreads 1 --customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000 --python_filename hlt_test_configs/TTbar_GEN_SIM_step.py
> ```
> 
> should the process modifier trigger the saving of the collection in the `FEVTDEBUG` content as well?

I assumed that the stubs were already saved in the L1 Trigger Event content, which was the case. However, their name seems to have changed from `l1tTkStubsGmt` to `l1tStubsGmt`. Updating the name in [L1Trigger/Configuration/L1Trigger_EventContent_cff.py](https://github.com/cms-sw/cmssw/blob/ba05945762fe45c9660349de58ba0db258243899/L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py#L254) solves the issue for me

Parsifal-2045 committed:
Update L1 Muon stubs naming to save them in FEVT

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46897/42973

- Found files with invalid states:
  - RecoMuon/L3TrackFinder/interface/phase2HLTMuonSelectorForL3.h:
    - Added: 9d7592fc16bdaaf222f91dd05c6eb512efe67f14
    - Deleted: 37b3b6fe38043a2acbdf41cceca9060df957b504
  - RecoMuon/L3TrackFinder/src/phase2HLTMuonSelectorForL3.cc:
    - Added: 9d7592fc16bdaaf222f91dd05c6eb512efe67f14
    - Deleted: 37b3b6fe38043a2acbdf41cceca9060df957b504

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #46453, #46853

cmsbuild commented:
Pull request #46897 was updated. @AdrianoDee, @Martin-Grunewald, @Moanwar, @aloeliger, @antoniovagnerini, @antoniovilela, @cmsbuild, @davidlange6, @dickychant, @epalencia, @fabiocos, @jfernan2, @mandrenguyen, @miquork, @mmusich, @rappoccio, @rseidita, @srimanob, @subirsarkar can you please check and sign again.

mmusich commented:
I am running other (private) tests in parallel -- in case you want to start main tests feel free to do it @Parsifal-2045

Parsifal-2045 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56c0a9/43387/summary.html
**COMMIT**: af19b588d9cc9e7f297cf07e930223bc91f0c2e1
**CMSSW**: CMSSW_15_0_X_2024-12-11-0800/el8_amd64_gcc12
**Additional Tests**: HLT_P2_INTEGRATION,HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46897/43387/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2024-12-11-0800&dataset=CMSSW_15_0_X_2024-12-11-0800-cmssw46897/el8_amd64_gcc12/61/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-11-0800+56c0a9/66075/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 453](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-11-0800+56c0a9/66075/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509544
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-11-0800+56c0a9/66075/triggerResults): found differences in 1 / 44 workflows

mmusich reviewed (commented):
I tested the branch running the following commands:
   * `hltPhase2UpgradeIntegrationTests --procModifiers phase2L2AndL3Muons`
   * `hltPhase2UpgradeIntegrationTests --procModifiers phase2L2AndL3Muons, phase2L3MuonsOIFirst`
   
and didn't observe any issues. I think we could move forward with this PR. I just leave few minor comments for my own education.

mmusich commented:
+hlt
   * for https://github.com/Parsifal-2045/cmssw/commit/af19b588d9cc9e7f297cf07e930223bc91f0c2e1
       * PR according to description and follow-up review.
   * branch passes phase-2 HLT central (and privately runs with modifiers) integration tests, see https://github.com/cms-sw/cmssw/pull/46897#pullrequestreview-2498916537
       * changes are active only upon usage of dedicated process modifiers.
   * I think the logic in `Phase2L2MuonSeedCreator` could be improved / simplified. OTOH this is not the final version of the code as per author description more updates are expected to recover full seeding efficiency. Let's start putting the code in cmssw to ease further development.

Parsifal-2045 commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,860 @@
+        // Found a matching segment in CSC -> propagate to ME2
+        LogDebug(metname) << "Found matching segment(s) in CSCs, propagating L1TkMu info to ME2 to seed";
+        // ME2
+        propagateToId = eta > 0 ? CSCDetId(1, 2, 0, 0, 0) : CSCDetId(2, 2, 0, 0, 0);
```
Second muon station in each endcap. As far as my understanding goes, the track-finding and track-fitting algorithms currently start propagating from ME2 so I've left the seed information propagating to that same station as in the current seeding module

Parsifal-2045 commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,860 @@
+        // Found at least one matching segment in DT -> propagate to MB2
+        LogDebug(metname) << "Found matching segment(s) in DTs, propagating L1TkMu info to MB2 to seed";
+        // MB2
+        propagateToId = DTChamberId(0, 2, 0);
```
Same as the endcap case, but for the second station in the barrel

Parsifal-2045 commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.h:
```diff
@@ -0,0 +1,152 @@
+  const double maxEtaOverlap_;  // overlap with |eta| < 1.3, endcap after that
+
+  // Handles
+  edm::ESHandle<MagneticField> magneticField_;
```
No reason, probably leftover from an older iteration, I'll remove it

Moanwar commented:
+Upgrade

Parsifal-2045 commented on RecoMuon/L2MuonSeedGenerator/src/Phase2L2MuonSeedCreator.cc:
```diff
@@ -0,0 +1,860 @@
+          container.push_back(dtSegments[matchingPair.first]);
+          matchedStations.push_back(detId.station());
+        }
+        for (int station = 1; station != 5; ++station) {
```
I found both 1 and 4 as static constants defined in the `DTChamberId` class (minimum station and maximum station in the barrel)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46897/43050

- Found files with invalid states:
  - RecoMuon/L3TrackFinder/interface/phase2HLTMuonSelectorForL3.h:
    - Added: 9d7592fc16bdaaf222f91dd05c6eb512efe67f14
    - Deleted: 62c4e649cc25c549d5a66ac804be1d00165335ab
  - RecoMuon/L3TrackFinder/src/phase2HLTMuonSelectorForL3.cc:
    - Added: 9d7592fc16bdaaf222f91dd05c6eb512efe67f14
    - Deleted: 62c4e649cc25c549d5a66ac804be1d00165335ab

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #46453, #46853, #46967

cmsbuild commented:
Pull request #46897 was updated. @AdrianoDee, @Martin-Grunewald, @aloeliger, @antoniovagnerini, @antoniovilela, @cmsbuild, @davidlange6, @dickychant, @epalencia, @fabiocos, @jfernan2, @mandrenguyen, @miquork, @mmusich, @rappoccio, @rseidita can you please check and sign again.

mmusich commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56c0a9/43495/summary.html
**COMMIT**: 93099c0fd3d77063fd9481ccc0648dda76196df7
**CMSSW**: CMSSW_15_0_X_2024-12-16-2300/el8_amd64_gcc12
**Additional Tests**: HLT_P2_INTEGRATION,HLT_P2_TIMING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46897/43495/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **HLT P2 Timing**: [chart](https://cmssdt.cern.ch/circles/web/piechart.php?data_name=hlt-p2-timing&resource=time_thread&filter=CMSSW_15_0_X_2024-12-16-2300&dataset=CMSSW_15_0_X_2024-12-16-2300-cmssw46897/el8_amd64_gcc12/63/Phase2Timing_resources)

## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-16-2300+56c0a9/66168/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-16-2300+56c0a9/66168/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 416](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-16-2300+56c0a9/66168/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509581
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-16-2300+56c0a9/66168/triggerResults): no differences found

mmusich commented:
+hlt
   * sign again, see https://github.com/cms-sw/cmssw/pull/46897#issuecomment-2547833729

jfernan2 commented:
+1

antoniovagnerini commented:
+dqm
- spurious differences from HGCal DQM test (open issue  #46416)

AdrianoDee commented:
+pdmv

mmusich commented:
@cms-sw/l1-l2 can you please take a look to the simple change in the list of keep statements of `L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py` ? 
Given the amount of signature involved in this PR, I'd like to avoid running into conflict.

aloeliger commented:
+l1

Looks fine from my end, thanks for catching the wrong "keep" statement because of the name change.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @sextonkennedy, @rappoccio, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
